crates/factor-outbound-pg/src/host.rs

```diff
@@ -0,0 +1,423 @@
+use anyhow::{anyhow, Result};
+use native_tls::TlsConnector;
+use postgres_native_tls::MakeTlsConnector;
+use spin_core::{async_trait, wasmtime::component::Resource};
+use spin_world::v1::postgres as v1;
+use spin_world::v1::rdbms_types as v1_types;
+use spin_world::v2::postgres::{self as v2, Connection};
+use spin_world::v2::rdbms_types;
+use spin_world::v2::rdbms_types::{Column, DbDataType, DbValue, ParameterValue, RowSet};
+use tokio_postgres::{
+    config::SslMode,
+    types::{ToSql, Type},
+    Client, NoTls, Row, Socket,
+};
+use tracing::instrument;
+use tracing::Level;
+
+use crate::InstanceState;
+
+impl InstanceState {
+    async fn open_connection(&mut self, address: &str) -> Result<Resource<Connection>, v2::Error> {
+        self.connections
+            .push(
+                build_client(address)
+                    .await
+                    .map_err(|e| v2::Error::ConnectionFailed(format!("{e:?}")))?,
+            )
+            .map_err(|_| v2::Error::ConnectionFailed("too many connections".into()))
+            .map(Resource::new_own)
+    }
+
+    async fn get_client(&mut self, connection: Resource<Connection>) -> Result<&Client, v2::Error> {
```

Comment:
This isn't blocking merging this PR since this is how things work in the current implementation, but it would be really nice if the host implementation here didn't assume a `tokio_postgres::Client`. 

Instead we could have the connection pool keep track of `Box<dyn PgClient>` where `PgClient` in this case is some trait that abstracts over the underlying postgres calls. 

When running this library as part of the wider Spin CLI, the `PgClient` impl would be backed by a `tokio_postgres::Client` but in test it would be backed by some dummy implementation.

This is what we're doing in the `llm` implementation so that tests don't need to run an actual LLM.  

Thoughts?

Reply:
:+1: It could also be a defaulted generic param on the whole factor (i.e. `OutboundPgFactor<Client = tokio_postgres::Client>`.

Reply:
Filed an issue to do this work in a followup PR #2649